File: memory-gateway/app/memory/engine.py
"""Deterministic memory engine pipeline (Phase 3 — no Memory AI required)."""

from __future__ import annotations

import logging
from dataclasses import dataclass, field

from app.memory.compressor import memory_ai_output_to_candidates
from app.memory.contradiction import ApplyResult
from app.memory.delta import DeltaResult
from app.memory.extractor import extract_candidates
from app.memory.low_info import is_low_info_message
from app.memory.state import (
    latest_context_version,
    list_memory_items,
    mark_items_obsolete,
    memory_changed,
    persist_candidates,
    write_context_version,
)
from app.models.memory import CandidateMemory, CanonicalMemorySnapshot, MemoryAIOutput, MemoryStatus
from app.providers.memory_ai import MemoryAIAdapter
from app.storage.db import session_scope

logger = logging.getLogger(__name__)


@dataclass
class MemoryUpdateResult:
    conversation_id: str
    skipped_low_info: bool = False
    candidates: int = 0
    applied: list[ApplyResult] = field(default_factory=list)
    obsolete_marked: int = 0
    context_version: int | None = None
    error: str | None = None

    @property
    def updated(self) -> bool:
        return memory_changed(self.applied, obsolete_count=self.obsolete_marked)


def process_memory_delta(
    delta: DeltaResult,
    *,
    memory_ai_output: MemoryAIOutput | None = None,
) -> MemoryUpdateResult | None:
    """
    Run memory update for a delta (deterministic + optional Memory AI).

    Fail-open: on any error return a result with error set (or None if
    session bootstrap fails); never raise into the proxy hot path.
    """
    if not delta.has_new:
        return MemoryUpdateResult(conversation_id=delta.conversation_id)

    # If every new user/assistant turn is low-info, skip extraction entirely.
    meaningful = [
        m
        for m in delta.new_messages
        if m.role in ("user", "assistant")
        and not is_low_info_message(m.content, role=m.role)
    ]
    if not meaningful:
        return MemoryUpdateResult(
            conversation_id=delta.conversation_id,
            skipped_low_info=True,
        )

    try:
        candidates: list[CandidateMemory] = extract_candidates(delta.new_messages)
        source_ids = [m.message_key for m in delta.new_messages]

        if memory_ai_output is not None:
            ai_candidates = memory_ai_output_to_candidates(memory_ai_output, source_ids)
            # Merge deterministic and AI candidates, avoiding exact duplicate content
            existing_contents = {c.content.casefold() for c in candidates}
            for ai_c in ai_candidates:
                if ai_c.content.casefold() not in existing_contents:
                    candidates.append(ai_c)
                    existing_contents.add(ai_c.content.casefold())

        with session_scope() as session:
            results: list[ApplyResult] = []
            if candidates:
                results = persist_candidates(session, delta.conversation_id, candidates)

            obsolete_count = 0
            if memory_ai_output is not None and memory_ai_output.obsolete_items:
                marked = mark_items_obsolete(
                    session,
                    delta.conversation_id,
                    memory_ai_output.obsolete_items,
                )
                obsolete_count = len(marked)

            version_num: int | None = None
            if memory_changed(results, obsolete_count=obsolete_count):
                cv = write_context_version(
                    session,
                    delta.conversation_id,
                    source_message_ids=source_ids,
                )
                version_num = cv.version
            return MemoryUpdateResult(
                conversation_id=delta.conversation_id,
                candidates=len(candidates),
                applied=results,
                obsolete_marked=obsolete_count,
                context_version=version_num,
            )
    except Exception as exc:
        logger.exception(
            "memory engine failed for conversation=%s; keeping prior memory",
            delta.conversation_id,
        )
        return MemoryUpdateResult(
            conversation_id=delta.conversation_id,
            error=str(exc),
        )


async def process_memory_delta_async(
    delta: DeltaResult,
    *,
    memory_ai: MemoryAIAdapter | None = None,
) -> MemoryUpdateResult | None:
    """
    Async pipeline calling Memory AI when available, falling back safely on error.
    """
    if not delta.has_new:
        return MemoryUpdateResult(conversation_id=delta.conversation_id)

    meaningful = [
        m
        for m in delta.new_messages
        if m.role in ("user", "assistant")
        and not is_low_info_message(m.content, role=m.role)
    ]
    if not meaningful:
        return MemoryUpdateResult(
            conversation_id=delta.conversation_id,
            skipped_low_info=True,
        )

    ai_output: MemoryAIOutput | None = None
    if memory_ai is not None:
        try:
            # Build delta messages text
            messages_text = "\n".join(
                f"{m.role}: {m.content}" for m in delta.new_messages if m.content
            )

            # Retrieve prior canonical snapshot summary if available
            prior_summary: str | None = None
            try:
                with session_scope() as session:
                    active = list_memory_items(
                        session,
                        delta.conversation_id,
                        status=MemoryStatus.ACTIVE.value,
                    )
                    if active:
                        snapshot = CanonicalMemorySnapshot.from_items(active)
                        prior_summary = snapshot.model_dump_json()
            except Exception:
                pass

            ai_output = await memory_ai.extract_memory(
                messages_text,
                current_memory_summary=prior_summary,
            )
        except Exception as exc:
            logger.warning(
                "Memory AI extraction failed for conversation=%s: %s; keeping prior memory/falling back to deterministic",
                delta.conversation_id,
                exc,
            )
            ai_output = None

    return process_memory_delta(delta, memory_ai_output=ai_output)

File: src/memory/engine.ts
import type { Database } from "../db";
import type { DeltaResult } from "./delta";
import type { ApplyResult } from "./contradiction";
import type { CandidateMemory, MemoryAIOutput } from "../models/memory";
import type { ShortTermContextStore } from "./context-store";
import { extractCandidates } from "./extractor";
import { isLowInfoMessage } from "./low-info";
import { analyzeCandidates } from "./analyzer";
import { memoryAiOutputToCandidates } from "./compressor";
import {
  listMemoryItems,
  markItemsObsolete,
  memoryChanged,
  persistCandidates,
  writeContextVersion,
} from "./state";
import { MemoryStatus, snapshotFromItems } from "../models/memory";
import type { MemoryAIAdapter } from "../providers/memory-ai";

export interface MemoryUpdateResult {
  userId: string;
  skippedLowInfo?: boolean;
  candidates: number;
  stored: number;
  contextCount: number;
  discarded: number;
  applied: ApplyResult[];
  obsoleteMarked: number;
  contextVersion?: number | null;
  error?: string | null;
}

export async function processMemoryDelta(
  db: Database,
  delta: DeltaResult,
  options?: {
    memoryAiOutput?: MemoryAIOutput | null;
    contextStore?: ShortTermContextStore | null;
  }
): Promise<MemoryUpdateResult | null> {
  const base = {
    userId: delta.userId,
    candidates: 0,
    stored: 0,
    contextCount: 0,
    discarded: 0,
    applied: [],
    obsoleteMarked: 0,
  };

  if (!delta.newMessages || delta.newMessages.length === 0) {
    return { ...base };
  }

  const meaningful = delta.newMessages.filter(
    (m) => ["user", "assistant"].includes(m.role) && !isLowInfoMessage(m.content, m.role)
  );

  if (meaningful.length === 0) {
    return {
      ...base,
      skippedLowInfo: true,
    };
  }

  try {
    const candidates: CandidateMemory[] = extractCandidates(delta.newMessages);
    const sourceIds = delta.newMessages.map((m) => m.messageKey);

    const memoryAiOutput = options?.memoryAiOutput;
    if (memoryAiOutput) {
      const aiCandidates = memoryAiOutputToCandidates(memoryAiOutput, sourceIds);
      const existingContents = new Set(candidates.map((c) => c.content.toLowerCase()));
      for (const aiC of aiCandidates) {
        if (!existingContents.has(aiC.content.toLowerCase())) {
          candidates.push(aiC);
          existingContents.add(aiC.content.toLowerCase());
        }
      }
    }

    // Memory Analyzer: route each candidate to store / context / discard.
    const { store, discard, contextEntries } = analyzeCandidates(candidates);

    let results: ApplyResult[] = [];
    if (store.length > 0) {
      results = await persistCandidates(db, delta.userId, store);
    }

    // Persist short-term context (Redis / in-memory store).
    let contextCount = 0;
    const contextStore = options?.contextStore;
    if (contextStore && contextEntries.length > 0) {
      for (const entry of contextEntries) {
        try {
          await contextStore.setContext(
            delta.userId,
            entry.key,
            entry.value,
            entry.ttlSeconds
          );
          contextCount++;
        } catch {
          // fail-open: context persistence must never break the main path
        }
      }
    }

    let obsoleteCount = 0;
    if (memoryAiOutput && memoryAiOutput.obsolete_items.length > 0) {
      const marked = await markItemsObsolete(
        db,
        delta.userId,
        memoryAiOutput.obsolete_items
      );
      obsoleteCount = marked.length;
    }

    let versionNum: number | null = null;
    if (memoryChanged(results, obsoleteCount)) {
      versionNum = await writeContextVersion(db, delta.userId, sourceIds);
    }

    return {
      userId: delta.userId,
      candidates: candidates.length,
      stored: store.length,
      contextCount,
      discarded: discard.length,
      applied: results,
      obsoleteMarked: obsoleteCount,
      contextVersion: versionNum,
    };
  } catch (err: any) {
    return {
      userId: delta.userId,
      candidates: 0,
      stored: 0,
      contextCount: 0,
      discarded: 0,
      applied: [],
      obsoleteMarked: 0,
      error: err?.message || String(err),
    };
  }
}

export async function processMemoryDeltaAsync(
  db: Database,
  delta: DeltaResult,
  options?: {
    memoryAi?: MemoryAIAdapter | null;
    contextStore?: ShortTermContextStore | null;
  }
): Promise<MemoryUpdateResult | null> {
  const base = {
    userId: delta.userId,
    candidates: 0,
    stored: 0,
    contextCount: 0,
    discarded: 0,
    applied: [],
    obsoleteMarked: 0,
  };

  if (!delta.newMessages || delta.newMessages.length === 0) {
    return { ...base };
  }

  const meaningful = delta.newMessages.filter(
    (m) => ["user", "assistant"].includes(m.role) && !isLowInfoMessage(m.content, m.role)
  );

  if (meaningful.length === 0) {
    return {
      ...base,
      skippedLowInfo: true,
    };
  }

  let aiOutput: MemoryAIOutput | null = null;
  const memoryAi = options?.memoryAi;

  if (memoryAi) {
    try {
      const messagesText = delta.newMessages
        .filter((m) => m.content)
        .map((m) => `${m.role}: ${m.content}`)
        .join("\n");

      let priorSummary: string | null = null;
      try {
        const active = await listMemoryItems(db, delta.userId, MemoryStatus.ACTIVE);
        if (active.length > 0) {
          priorSummary = JSON.stringify(snapshotFromItems(active));
        }
      } catch {}

      aiOutput = await memoryAi.extractMemory(messagesText, priorSummary);
    } catch {
      aiOutput = null;
    }
  }

  return await processMemoryDelta(db, delta, {
    memoryAiOutput: aiOutput,
    contextStore: options?.contextStore,
  });
}
